### PR TITLE
Denoise merge

### PR DESCRIPTION
**Pull request by cheesema**
*Introduces by level learning of de-noising kernels using iterative linear solvers on local patches.

### CODE REVIEW

File: examples/Example_denoise.hpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    // APR datastructure
+    APR apr;
+
+    timer.start_timer("read apr");
+    //read file
+    APRFile aprFile;
+    aprFile.open(file_name,"READ");
+    aprFile.read_apr(apr);
+
+    ParticleData<float>parts;
```
Comment (joeljonsson) [inline examples/Example_denoise.hpp:66]: Should we make this uint16_t, since this is what is produced by the other examples by default?

**Comment by psteinb** [comment]
Just wondering: what is the conceptual approach to denoising taken here?

**Comment by cheesema** [comment]
Very good quesiont @psteinb. It is an idea that I started playing with 2 or so years ago, in another context. It is similar to N2V like codes but much simpler.

You learn local kernels (one layer) say 5x5x5, where you train them predicting the center pixel. You do this where you have one kernel per layer of the APR. So some of the complexity is handled direclty by the APR already. The kernels are trained in an iterative fashion, with random sampling used over local patches and an average solution found. It is unsupervised. The current implimentation can be optimised (w.r.t) performance as it relies on local reconstructions, rather then using some of the local filter code that Joel has recently introduced. 

I would have to refresh myself on some of the finer details of the iterative optimisation scheme. It is easy to train, and seems to work fairly well. Although would love to do some more quantitative comparisons against the current state of the art.

**Review by joeljonsson — COMMENTED** [review]
This seems to be working now, but there are a few things that should be done:
- [ ] Filter application should use the existing APRFilter implementation
- [ ] Add option to specify the filter size, and possibly reduce filter size with decreasing level
- [ ] Remove the (currently unused) nonlinear filters OR fix them up and put them in separate functions / add a flag

File: examples/Example_denoise.hpp (hunk 1 above)
Comment (cheesema) [reply examples/Example_denoise.hpp]: yes.

**Comment by cheesema** [comment]
Agreed on point 1. Point 3 is done, point 2 that is effectively there. Although should be made easier and consistent with the APRFIlter implimentation in point 1.

**Comment by cheesema** [comment]
Shall we create some tasks to track here in Github for those remaining?

**Comment by cheesema** [comment]
Also the training code needs to be cleaned up. It is pretty awful.

**Comment by cheesema** [comment]
In light of Peters comments, should also write up a short technical note on what it is doing.

**Review by joeljonsson — APPROVED** [review]
Since this is working, I'm approving this for now, and we'll add in the improvements later

**Pull request merged**